feat: Use hatchling as the build backend

matthewfeickert opened the pull request:
To be able to achieve https://github.com/ssl-hep/ServiceX_frontend/issues/352 all dependencies need to be on conda-forge as well. conda-forge requires that all distributions also package their LICENSE file, and conda-forge builds the distributions from either Git source code or (preferably) a sdist on PyPI. `make-it-sync`'s sdist doesn't include the LICENSE file

```console
$ python -m pip download --no-binary make-it-sync make-it-sync
Collecting make-it-sync
  Downloading make-it-sync-1.0.0.tar.gz (3.5 kB)
  Preparing metadata (setup.py) ... done
Saved ./make-it-sync-1.0.0.tar.gz
Successfully downloaded make-it-sync
$ python -m tarfile --list make-it-sync-1.0.0.tar.gz 
make-it-sync-1.0.0/ 
make-it-sync-1.0.0/PKG-INFO 
make-it-sync-1.0.0/README.md 
make-it-sync-1.0.0/make_it_sync/ 
make-it-sync-1.0.0/make_it_sync/__init__.py 
make-it-sync-1.0.0/make_it_sync/func_wrapper.py 
make-it-sync-1.0.0/make_it_sync.egg-info/ 
make-it-sync-1.0.0/make_it_sync.egg-info/PKG-INFO 
make-it-sync-1.0.0/make_it_sync.egg-info/SOURCES.txt 
make-it-sync-1.0.0/make_it_sync.egg-info/dependency_links.txt 
make-it-sync-1.0.0/make_it_sync.egg-info/requires.txt 
make-it-sync-1.0.0/make_it_sync.egg-info/top_level.txt 
make-it-sync-1.0.0/setup.cfg 
make-it-sync-1.0.0/setup.py 
```

As this requires making a new release that includes LICENSE information, I updated the build backend from `setuptools`+`distutils` to `hatchling`.

@gordonwatts If this all looks acceptable, after this is merged could you make a `v0.2.0` (minor as this drops Python 3.6 as it is very EOL) release on PyPI?

matthewfeickert committed:
feat: Use hatchling as the build backend

* Migrate from setuptools to hatchling for the build backend.
* Support Python 3.7+.
* Remove setup.py.

matthewfeickert committed:
ci: Update CI workflows to support testing Python 3.12

gordonwatts commented:
I've never used hatch - it is like there is a new tool every 10 days and I don't know how you keep switching Matthew - it feels like it takes too much time! But this is a great upgrade - a toml file! Thank you so much!!!!!

matthewfeickert commented:
> I've never used hatch - it is like there is a new tool every 10 days and I don't know how you keep switching

I switched once when we moved to `hatch`/`hatchling` for `pyhf` and I'm staying until I need compiled dependencies or someone gives me a very good argument to change! :P

matthewfeickert commented:
Also thanks for being super speedy with a [`v2.0.0` release](https://github.com/gordonwatts/make-it-sync/releases/tag/2.0.0)! :rocket:
